### DISCUSSION TITLE
Any ideas to make navigation-tree touch-friendly?

### DISCUSSION THREAD

**Discussion by msharp1** [datenstrom/community#123] [Ask a question] [answered]
Hi,
I've noticed that when I using navigation-tree that I'm not able to tap sub-entry.
I'm not sure if flatsite theme is aimed to be fully responsive ?

**Comment by markseuffert** [comment]
Good idea to make all themes touch-friendly. Can you describe what's not working with the flatsite theme? Do other themes work for you, e.g. blogsite theme? Any help would be much appreciated. :smile:

**Comment by msharp1** [comment]
Hi Mark,
As mentioned I'm speaking specifically of 'navigation-tree' feature, which is unusable from a touch device perspective. Either on flatsite or blogsite. The dropdown menu open when you touch the link label (ex: 'Wiki') but then you are redirected to the page corresponding to this link without a chance to select any sub-entry. That's a common touch-related issue that could be solved by using extended CSS or JQuery code...Or using a framework that provide that kind of components. I'm not a programmer, but as far as I can see on the internet there is not really a unique good way to achieve that, but several ones...

**Comment by msharp1** [comment]
See UIKit CSS framework for example : http://getuikit.com/docs/documentation_get-started.html

**Comment by markseuffert** [comment]
As far as I know, the dropdown menu works on IOS. Other touch devices don't support a pure CSS dropdown menu. They need JavaScript or jQuery to toggle the submenu between hidden and visible. Let us know if someone finds a good solution, contributions are always welcome. 

Here's another example from [Pure](http://purecss.io/menus/#dropdowns).

**Comment by msharp1** [comment]
The last example from Pure, with the "contact" dropdown works on my Android phone. (but there is a javascript in behind).
The one at http://codepen.io/andornagy/pen/xhiJH which seems 100% CSS seems to work too.

The most I can say about those menus in general, and especially between IOS/Android/Desktop is this is really a mess and really not straight forward. I spent time on googling yesterday and there is no real best practice about it, everyone having their own mixed ways to do it... We need to choose something and indeed a real modern web guru could help...

Maybe http://osvaldas.info/drop-down-navigation-responsive-and-touch-friendly ?

**Comment by msharp1** [comment]
What is good with the one at http://osvaldas.info/drop-down-navigation-responsive-and-touch-friendly is it stacks in hamburger style on mobile  ;)

**Comment by dieli** [comment]
Is the navigation on http://www.dielacher.org/ something that works for you/is what you want? It's pure CSS.

**Comment by msharp1** [comment]
Hi @dieli, and thank you for taking part of the conversation :)
Yes, it works on my Android phone. The only little glitch is the submenu shows up inevitably unfolded but at least it makes sub-entries available. You don't want to have too many sub-entries with it.

PS : note that it is not specifically what **I want**, it is what every Android touch device user would need (at least) to access sub-entries.

Cheers,

**Comment by dieli** [comment]
@calimou you can check out my theme at https://github.com/dieli/yellow-theme-pure. Maybe you find a solution :-)

**Comment by msharp1** [comment]
Thanks.
Basically, I have in mind that Yellow should be mobile-ready from scratch, because those days a lot of people are using touch devices. I would say it's vital for the CMS life and maybe that should be a priority because we live a modern time ;)
That being said, I'll try to implement something based on the suggestion, thanks again !

**Comment by dominikwilkowski** [comment]
Hey guys,

You can make a dropdown with pure CSS but it must include a hidden checkbox which is not very semantic. If you care about that(which you should) go with a simple vanilla js event-driven script that just toggles classes for the dropdowns. (http://gel.westpacgroup.com.au/newsite/ shows you how but a prototype that will be moved soon)

For a look at a pure CSS driven nav, have a look at this: http://codepen.io/dominikwilkowski/pen/avjeGb

I just wrote something very quick on the bus so don't hunt me down for poor design ;) Just a quick mockup.

The problem is more the non-existent hover-state on touch devices. iOS tried to "fix" that by adding a second tab to every clickable element(eg link) that also has an event handler attached to it. That is very unintuitive and just slows everyone down. The only way to get around this afaik is to have a link that is purely for opening the dropdown and then inside that dropdown have a link right on the top that links to the "category-page". That approach is visible on http://www.asic.gov.au which I wrote a while back. The link for the main menu is clickable but repeated within the dropdown itself again. (It also has an animation delay on it to make sure the intend is there to open the menu).

Hope that helps at all after I just skim-read the thread :)

**Comment by msharp1** [comment]
Thanks @dominikwilkowski,
Sooo, our options are : 
- 100% CSS but with CSS/HTML checkbox hacking (not very semantic)
- Not 100% CSS but more semantic - with javascript (I don't clearly understand where is that code when looking at the source of http://gel.westpacgroup.com.au/newsite/)

About http://codepen.io/dominikwilkowski/pen/avjeGb vs http://www.asic.gov.au/ I see behavior differences  : at http://www.asic.gov.au/ the dropdown is **both** a link and a dropdown trigger.

**Comment by dominikwilkowski** [comment, marked as answer]
There are several ways.

Sorry guys I was a bit short on time last night(Australia time here)

To sumerize:
- Dropdown with checkbox  
  http://codepen.io/dominikwilkowski/pen/avjeGb
- Dropdown with focus  
  http://codepen.io/dominikwilkowski/pen/avamZx
- Dropdown with js classes  
  http://codepen.io/dominikwilkowski/pen/bVxwgr

Now those are very high-level. You can now add delays to make sure the user-intent is there. You can work on more details like removing the class in the js example when you click somewhere else and the list goes on.

To me it is very important to make the dropdowns accessible. We got a nested list here which is pretty accessible already to a screen reader but it also needs to be keyboard accessible. So in the js example you would need to look for focus, enter as well as click. Also it falls back to a normal open dropdown when js is broken or unavailable. That too is an accessibility concern.

But to your actual concern: all of them should work in touch devices with the notable exception of the focus one. Focus and hover is just convoluted on touch so just assume you ain't got none ;)

To make something really stable and usable you would probably use modernizr and change behaviour depending wether you are on touch or not. (That's what we did on asic.

Sidenote: having a sneaky look at the [demo](http://demo.datenstrom.se/) I see that you guys are not HTML5 yet? Any reasons? Also not AA accessible and whatnot :)

**Comment by markseuffert** [comment]
Thank you Dominik, appreciate your help. :smile: 

I can confirm the first and third example is working on IOS. Anyone can test on Android?

**Comment by msharp1** [comment]
Thanks a lot Dominik. I've tested on my Android phone, 1 & 3 examples works with the following remarks:
- on ex3, dropdown section label can hold a link. (ex1 can't ?).
- on ex3 I'm not able to collapse the dropdown while on ex1 it's still possible by clicking again on the section label.

@markseu , @dominikwilkowski : the notice about HTML5 compliance could justify a new issue request ?

**Comment by dominikwilkowski** [comment]
There is a lot you can do once you decide to go with js... (ex3).
Ex1 can contain a link.. and it does. I think I am missing something here :)

We got a testing lab here at work with all kinds of droids and windows phones and other redundant tech :) They all work

**Comment by msharp1** [comment]
Hi @dominikwilkowski,
Sorry if I'm unclear or I don't understand something: let me rephrase (it's more a question) : 

The menu structure should be as follow  (or not, but we should decide !): 
- Top switch (section) => **link**
  - sub entry => link
  - sub entry => link...

And I'm speaking about the first level link in bold above.

With the following code (ex1), which is responsible for the switch and what I call the "dropdown section label", can we include a link so it opens a page ?

```
<label for="navigation__switch1">Ideas</label>
<input class="navigation__switch" type="checkbox" id="navigation__switch1">
```

EDIT - I suppose just by adding a `<a href=""></a>` around the label :)

```
<a href="#section-link"><label for="navigation__switch1">Ideas</label></a>
<input class="navigation__switch" type="checkbox" id="navigation__switch1">
```

Sorry for the newbie noise :)

**Comment by dominikwilkowski** [comment]
Well let me ask you this: How do you see this working on a touch device?
If the first level link is an actual link that goes to a page, what will happen on a touch device when you touch it? Open the dropdown or go to the link?

PS you can't nest a link and a label. Both have click events. The browser wouldn't know what to do :)

**Comment by msharp1** [comment]
> Open the dropdown or go to the link?

Indeed you're perfectly right ! And this has to be decided : with the current navigation-tree implementation, on a desktop, the firstlevel link is clickable and opens the corresponding page.
For touch devices, I suggest it only opens the dropdown and display the firstlevel as an _additional_ link, like you mentioned before - so we still can access the firstlevel's corresponding page.

We also have the "DoubleTapToGo.js" option from the [website I mentioned before](http://osvaldas.info/drop-down-navigation-responsive-and-touch-friendly) which is rather "advanced" :D 

On very small screen, the better would be the 'hamburger' mode.

Now, who's gonna implement this for Yellow ? What do you think @markseu  ? I can offer my testing :)

**Comment by dominikwilkowski** [comment]
I also suggest a hamburger approach. Morph the nav into a flyout menu and show a hamburger. With js all you have to do then is to add a class to the body for open flyout and the rest is done in CSS. Clean and works in all devices and browsers.

E.g. a website I did: https://www.stgeorge.com.au/ with this

**Comment by markseuffert** [comment]
> Now, who's gonna implement this for Yellow? 

I suggest that someone who needs this features implements it. Any help would be much appreciated. :smile: 

If others would like to see this feature, please add +1 to this issue.
